libraries/Networking/src/NetworkInterface.cpp

```diff
@@ -0,0 +1,719 @@
+        _ip_ev_instance = NULL;
+    }
+    if(_esp_netif != NULL){
+        esp_netif_destroy(_esp_netif);
+        _esp_netif = NULL;
+    }
+    if(_interface_event_group != NULL){
+        vEventGroupDelete(_interface_event_group);
+        _interface_event_group = NULL;
+        _initial_bits = 0;
+    }
+}
+
+bool NetworkInterface::initNetif(Network_Interface_ID interface_id, bool server_interface) {
+    if(_esp_netif == NULL || interface_id >= ESP_NETIF_ID_MAX){
+        return false;
+    }
+    _is_server_if = server_interface;
+    _interface_id = interface_id;
+    _got_ip_event_id = esp_netif_get_event_id(_esp_netif, ESP_NETIF_IP_EVENT_GOT_IP);
+    _lost_ip_event_id = esp_netif_get_event_id(_esp_netif, ESP_NETIF_IP_EVENT_LOST_IP);
+    _interfaces[_interface_id] = this;
+
+    if(_interface_event_group == NULL){
+        _interface_event_group = xEventGroupCreate();
+        if(!_interface_event_group){
+            log_e("Interface Event Group Create Failed!");
+            return false;
+        }
+        setStatusBits(_initial_bits);
+    }
+
+    if(_ip_ev_instance == NULL && esp_event_handler_instance_register(IP_EVENT, ESP_EVENT_ANY_ID, &_ip_event_cb, NULL, &_ip_ev_instance)){
+        log_e("event_handler_instance_register for IP_EVENT Failed!");
+        return false;
+    }
+    return true;
+}
+
+bool NetworkInterface::started() const {
+    return (getStatusBits() & ESP_NETIF_STARTED_BIT) != 0;
+}
+
+bool NetworkInterface::connected() const {
+    return (getStatusBits() & ESP_NETIF_CONNECTED_BIT) != 0;
+}
+
+bool NetworkInterface::hasIP() const {
+    return (getStatusBits() & ESP_NETIF_HAS_IP_BIT) != 0;
+}
+
+bool NetworkInterface::hasLinkLocalIPv6() const {
+    return (getStatusBits() & ESP_NETIF_HAS_LOCAL_IP6_BIT) != 0;
+}
+
+bool NetworkInterface::hasGlobalIPv6() const {
+    return (getStatusBits() & ESP_NETIF_HAS_GLOBAL_IP6_BIT) != 0;
+}
+
+bool NetworkInterface::enableIPv6(bool en)
+{
+    if (en) {
+        setStatusBits(ESP_NETIF_WANT_IP6_BIT);
+    } else {
+        clearStatusBits(ESP_NETIF_WANT_IP6_BIT);
+    }
+    return true;
+}
+
+bool NetworkInterface::dnsIP(uint8_t dns_no, IPAddress ip)
+{
+    if(_esp_netif == NULL || dns_no > 2){
+        return false;
+    }
+    if(_is_server_if && dns_no > 0){
+        log_e("Server interfaces can have only one DNS server.");
+        return false;
+    }
+    esp_netif_dns_info_t d;
+    // ToDo: can this work with IPv6 addresses?
+    d.ip.type = IPADDR_TYPE_V4;
+    if(static_cast<uint32_t>(ip) != 0){
+        d.ip.u_addr.ip4.addr = static_cast<uint32_t>(ip);
+    } else {
+        d.ip.u_addr.ip4.addr = 0;
+    }
+    if(esp_netif_set_dns_info(_esp_netif, (esp_netif_dns_type_t)dns_no, &d) != ESP_OK){
+        return false;
+    }
+    return true;
+}
+
+bool NetworkInterface::config(IPAddress local_ip, IPAddress gateway, IPAddress subnet, IPAddress dns1, IPAddress dns2, IPAddress dns3)
```

Comment:
The notes in the header say `// For server interfaces (WiFi AP), dns1 is the DHCP lease range start and dns2 is the DNS. dns3 is not used`.

That is a bit weird, having the fourth parameter be the DHCP range, and only one DNS.

Maybe two separate functions for `config()` [normal] and `configServer(local_ip, gateway, subnet, dhcp_range, dns)`.  The functions could check `_is_server_if` and return an error / failure if it is not the right type (e.g. call server config if not a server).

Reply:
it's more than we have now. We can think of something less weird later. I don't even guarantee that the DNS server is properly advertised by the DHCPS